_source/_posts/2022-03-15-programming-security-and-why-rust.md

```diff
@@ -0,0 +1,157 @@
+---
+layout: blog_post
+title: "Why Safe Programming Matters and Why a Language Like Rust Matters"
+author: deepu-sasidharan
+by: advocate
+communities: [security, rust]
+description: "Why safe programming matters and why I would choose Rust for that. We will also talk about Fast and safe programming with Rust."
+tags: [security, rust]
+tweets:
+  - "Why does safe programming matter? Learn what safe programming is and why Rust is a great choice. #rustlang #rust #programming #security"
+  - "Are you still writing C/C++ code? You will be amazed at the amount issues arising from unsafe code. #rustlang #rust #programming #security"
+image: blog/programming-security-and-why-rust/cover.jpg
+type: awareness
+---
+
+As programmers, how many of you have a good understanding of programming safety or secure programming? It's not the same application security or cyber security. I have to confess; I didn't know a lot about these in the early years of my career, especially since I didn't come from a computer science background. But looking back, I think programming security is something every programmer should be aware of and should be taught at a junior level.
+
+What is safe programming, or to be more precise, what does being safe mean for a programming language? or rather what does unsafe mean? Let's set the context first.
+
+{% include toc.md %}
+
+# Programming safety
+
+> Programming Safety = Memory safety + Type safety + Thread safety
+
+Safety can be categorized into three, memory safety, type safety, and thread safety (or four if you count null safety as a separate item)
+
+## Memory safety
+
```

Comment:
```suggestion
If you've ever had a program run out of memory due to a memory leak or accidentally read from an index of an array that was greater than the array's length, you've suffered from lack of memory safety.
```

Reply:
this is a bit too generic IMO since most memory-safe languages including Rust can have memory leaks and its not unsafe perse. Same for array out of bounds. These become memory safety issue only in C/C++